framework/src/mediaplayer/mediaplayer.c

```diff
@@ -0,0 +1,581 @@
+{
+	pthread_mutex_lock(&g_mutex_playing_q);
+	dq_addlast((dq_entry_t *)m, &g_playing_q);
+	pthread_mutex_unlock(&g_mutex_playing_q);
+	sem_post(&g_sem_playing);
+	return MEDIA_OK;
+}
+
+media_result_t media_record(media_t *m)
+{
+#ifdef CONFIG_AUDIO_MULTI_CARD
+	pthread_mutex_lock(&g_mutex_recording_q);
+	dq_addlast((dq_entry_t *)m, &g_recording_q);
+	pthread_mutex_unlock(&g_mutex_recording_q);
+	sem_post(&g_sem_recording);
+#else
+	pthread_mutex_lock(&g_mutex_playing_q);
+	dq_addlast((dq_entry_t *)m, &g_playing_q);
+	pthread_mutex_unlock(&g_mutex_playing_q);
+	sem_post(&g_sem_playing);
+#endif
+	return MEDIA_OK;
+}
+
+media_result_t media_stop(media_t *m)
+{
+	if (m->state == MEDIA_STATE_PLAYING || m->state == MEDIA_STATE_RECORDING) {
+		m->state = MEDIA_STATE_STOPPING;
+		return MEDIA_OK;
+	}  
+	return MEDIA_ERROR;
+}
+
+media_result_t media_pause(media_t *m)
+{
+	if (m->state == MEDIA_STATE_PLAYING) {
+		m->state = MEDIA_STATE_PAUSING;
+		return MEDIA_OK;
+	}
+	return MEDIA_ERROR;
+}
+
+media_result_t media_resume(media_t *m)
+{
+	if (m->state == MEDIA_STATE_PAUSED) {
+		m->state = MEDIA_STATE_PLAYING;
+		sem_post(&g_sem_playing);
+		return MEDIA_OK;
+	}
+	return MEDIA_ERROR;
+}
+
+void media_setvol(media_t *m, unsigned char vol)
+{
+	return;
+}
+
+bool media_isplaying(media_t *m)
+{
+	return (m->state == MEDIA_STATE_PLAYING);
+}
+
+media_result_t media_getvol(media_t *m)
+{
+	return MEDIA_ERROR;
+}
+
+media_result_t media_seek(media_t *m, int time_ms)
+{
+	return MEDIA_OK;
+}
+
+media_result_t media_setposition(media_t *m, int time_ms)
+{
+	return MEDIA_OK;
+}
+
+media_result_t media_getposition(media_t *m)
+{
+	return MEDIA_OK;
+}
+
+/***
+ * Functions run on threads
+ */
+
+static media_result_t read_wav_header(media_t *m)
+{
+	//struct pcm_config config;
+	struct wav_header_s header;
+	int ret;
+
+	ret = read(m->fd, &header, sizeof(header));
+	if (ret != sizeof(header)) {
+		return MEDIA_ERROR_UNKNOWN_FILE;
+	}
+
+	// Todo: Add header information to media_t
+
+	return MEDIA_OK;
+}
+
+media_t *media_open(char *path, media_op_t op, media_type_t type, media_play_type_t play_type)
+{
+	media_t *m = (media_t *)malloc(sizeof(media_t));
+	if (m == NULL) {
+		return NULL;
+	}
+
+	m->fd = -1;
+	m->op = op;
+	m->state = MEDIA_STATE_CREATED;
+	m->type = type;
+	m->play_type = play_type;
+
+	if (type == MEDIA_TYPE_WAV || type == MEDIA_TYPE_PCM) {
+		if (op == MEDIA_OP_PLAYBACK) {
+			if (play_type == MEDIA_PLAY_TYPE_FILE) {
+				m->fd = open(path, O_RDONLY);
+				if (m->fd < 0) {
+					free(m);
+					return NULL;
+				}
+
+				if (m->type == MEDIA_TYPE_WAV) {
+					read_wav_header(m);
+				}
+				m->state = MEDIA_STATE_PLAYING;
+				return m;
+			} else if (play_type == MEDIA_PLAY_TYPE_NETWORK) {
+				// TODO: Network streaming support
+				free(m);
+				return NULL;
+			}
+			free(m);
+			return NULL;
+
+		} else if (op == MEDIA_OP_RECORD) {
+			if (play_type == MEDIA_PLAY_TYPE_FILE) {
+				m->fd = open(path, O_RDWR | O_CREAT | O_TRUNC);
```

Comment:
Ok, now I understand your last reply.
Then I think we'd better use fd value instead of path.
If we use O_TRUNC here, it always remove old file forcely.
Let's user decide it.

Reply:
User will never open the file directly in the current implementation. He/She will pass filename . If file open fails, then NULL will be returned. I will add some debug statements here.

Reply:
What I'm saying is that user should decide O_APPEND or O_TRUNC.
Currently user never append voice data in same file.